Automatic symmetry detection using spglib

jaemolihm opened the pull request:
* Added method `System.set_structure` that sets `positions`, `atom_numbers` (Integer numbers to distinguish species), and `magnetic_moments` (optional) attributes.
* Added method `System.set_symmetry_from_structure` that uses spglib to automatically detect symmetry of the system and sets the `symgroup` attribute.

Implements the first two steps of #139.

Assumes #140, #146 are merged.

Fixes #144.

For magnetic systems, symmetries involving time reversal (e.g. `SYM.C2x * SYM.TimeReversal`) are not detected because spglib does not support such symmetries. (For non-magnetic systems, TimeReversal is added to the generator so there is no problem.)

### TODO
* [x] Magnetic symmetry
* [x] Manual documentation in website? : Docstring is already generated.
* [x] More tests

stepan-tsirkin commented on wannierberri/__system.py:
```diff
@@ -409,6 +409,66 @@ def check_hermitian(self,XX):
+        ----------
+        positions : (num_atom, 3) array_like of float
+            Atomic positions in fractional coordinates.
+        atom_numbers: (num_atom,) list of int
```
would be good to provide `atom_labels` -  a list of any labels (string,number, or in general any hashable type, e.g.`['Fe','O1','O2','O1',2,3.14]` and then internally map them to numbers to pass to the `spglib`

stepan-tsirkin commented on wannierberri/__system.py:
```diff
@@ -409,6 +409,66 @@ def check_hermitian(self,XX):
+        if self.magnetic_moments is None:
+            symmetry_gen.append(TimeReversal)
+        else:
+            print("Warning: spglib does not support time reversal symmetry for noncollinear "
```
@jaemolihm , did you try testing `spglib` mith magnetic moments ? does it work?

I tried and it looks to neglect them, e.g. 

```
In [3]: spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]],[ [-0.1,-0.1,-0.1],[0.1,0.1,0.1]],[1,1],[-1,1]))                                                                                                                                                                                         
Out[3]: 
{'rotations': array([[[ 1,  0,  0],
         [ 0,  1,  0],
         [ 0,  0,  1]],
 
        [[-1,  0,  0],
         [ 0, -1,  0],
         [ 0,  0, -1]]], dtype=int32),
 'translations': array([[0., 0., 0.],
        [0., 0., 0.]]),
 'equivalent_atoms': array([0, 0], dtype=int32),
 'primitive_lattice': array([[1., 0., 0.],
        [0., 2., 0.],
        [0., 0., 3.]])}

In [4]: spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]],[ [-0.1,-0.1,-0.1],[0.1,0.1,0.1]],[1,1],[1,1]))                                                                                                                                                                                          
Out[4]: 
{'rotations': array([[[ 1,  0,  0],
         [ 0,  1,  0],
         [ 0,  0,  1]],
 
        [[-1,  0,  0],
         [ 0, -1,  0],
         [ 0,  0, -1]]], dtype=int32),
 'translations': array([[0., 0., 0.],
        [0., 0., 0.]]),
 'equivalent_atoms': array([0, 0], dtype=int32),
 'primitive_lattice': array([[1., 0., 0.],
        [0., 2., 0.],
        [0., 0., 3.]])}

```

Result is unaffected by the fourth entry of `cell`

jaemolihm replied:
You are right. `spglib` does not work with magnetic moments.

stepan-tsirkin replied:
then, maybe, for this PR just disable magnetic moments, and write in the docs that only TRS systems are allowed. implement magnetic systems later.

jaemolihm replied:
Sorry, I misunderstood what you meant.

I think your are correct that spglib does not work for collinear magnetic moments as input. (It says it works, so maybe I am using wrong input.)

Spglib does work for noncollinear magnetic moments as input. For collinear systems, one can set the magnetic moments to e.g. [0., 0., 1.]. For example (based on your code, but changed only the magnetic moments argument to a vector:
```python3
>>> spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]], [ [-0.1,-0.1,-0.1], [0.1,0.1,0.1]], [1,1], [[0.,0.,1.], [0.,0.,1.]]))
{'rotations': array([[[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]]], dtype=int32), 'translations': array([[0., 0., 0.],
       [0., 0., 0.]]), 'equivalent_atoms': array([0, 0], dtype=int32), 'primitive_lattice': array([[1., 0., 0.],
       [0., 2., 0.],
       [0., 0., 3.]])}
>>> spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]], [ [-0.1,-0.1,-0.1], [0.1,0.1,0.1]], [1,1], [[0.,0.,1.], [0.,0.,-1.]]))
{'rotations': array([[[1, 0, 0],
        [0, 1, 0],
        [0, 0, 1]]], dtype=int32), 'translations': array([[0., 0., 0.]]), 'equivalent_atoms': array([0, 1], dtype=int32), 'primitive_lattice': array([[1., 0., 0.],
       [0., 2., 0.],
       [0., 0., 3.]])}
```

The limitation in this case is that spglib misses out the symmetry operations that involve time reversal (such as  time reversal * C2z). In short, spglib works but finds only a subset of the actual symmetry of the system.

The case with magnetic moments is actually being tested for Fe: see test_symmetry_spglib_Fe.

What I can do is to add an error if `magnetic_moments` is set to a number, not a vector. I will do this in this PR soon.

stepan-tsirkin replied:
> What I can do is to raise an error if magnetic_moments is set to a list of numbers, not a list of vectors.

I think it is a way to go, not to complicate things at this point

> Or, one can convert from numbers to vectors internally

 in that case one needs to specify the quantization axis (which is not always `z`, but may have a default value `[0,0,1]`

>  The problem is that spglib does not tell you whether the given symmetry operation involves time reversal or not

Then it is (almost) useless information for us. Well, for the TR-even tensors it may be used. But I suggest to consider only the non-collinear case. Anyway, in a collinear case the spin-up and spin-down subsystems may be considered independently. 

So, just make sure magnetic moments passed to spglib are vectors not numbers, either way.

stepan-tsirkin commented on wannierberri/__system.py:
```diff
@@ -409,6 +409,66 @@ def check_hermitian(self,XX):
+            print("Warning: spglib does not support time reversal symmetry for noncollinear "
+                  "systems.\nTo include such symmetries, use set_symmetry.")
+
+        self.symgroup = Group(symmetry_gen, recip_lattice=self.recip_lattice, real_lattice=self.real_lattice)
```
A good way to go would be to determine all symmetries without account of magmom, and then try if the symmetries found do preserve magmom, either alone or in combination with TR

jaemolihm replied:
I agree, it seems like a good idea.

stepan-tsirkin replied:
>  I was a bit reluctant to implement that because ideally that code should be in spglib.
Then go ahead and contribute to `spglib` !

jaemolihm commented:
I updated the code. I used `atom_labels` instead of `atom_numbers`, and added an error if `magnetic_moments` is set to a list of numbers, not a list of 3d vectors.
